Added `.pop_child`, `.pop_related`, `.despawn_first_related`, `.despawn_last_related`

MuongKimhong opened the pull request:
# Objective

- Remove relationship between an entity and its last child.
- Despawn last child of an entity.
- Despawn first child of an entity.

## Solution

#### Added below methods to `EntityWorldMut` and `EntityCommands`.
- `pop_child()`
- `pop_related()`
- `despawn_first_related()`
- `despawn_last_related()`

## Testing
```rust
#[test]
fn pop_child() {
    let mut world = World::new();
    let child1 = world.spawn_empty().id();
    let child2 = world.spawn_empty().id();

    let mut root = world.spawn_empty();
    root.add_children(&[child1, child2]);
    root.pop_child();
    let root = root.id();

    let hierarchy = get_hierarchy(&world, root);
    assert_eq!(
        hierarchy,
        Node::new_with(root, vec![Node::new(child1)])
    );
}

#[test]
fn test_pop_related() {
    let mut world = World::new();

    let a = world.spawn_empty().id();
    let b = world.spawn(ChildOf(a)).id();
    let c = world.spawn(ChildOf(a)).id();
    let d = world.spawn(ChildOf(a)).id();

    world.entity_mut(a).pop_related::<Children>();

    assert_eq!(world.entity(d).get::<ChildOf>(), None);
    assert!(world.entity(b).get::<ChildOf>().is_some());
    assert!(world.entity(c).get::<ChildOf>().is_some());
}

#[test]
fn test_despawn_first_and_last_related() {
    let mut world = World::new();

    let a = world.spawn_empty().id();
    let b = world.spawn(ChildOf(a)).id();
    let c = world.spawn(ChildOf(a)).id();
    let d = world.spawn(ChildOf(a)).id();

    world.entity_mut(a).despawn_first_related::<Children>();
    world.entity_mut(a).despawn_last_related::<Children>();

    assert!(world.get_entity(b).is_err());
    assert!(world.get_entity(c).is_ok());
    assert!(world.get_entity(d).is_err());
}
```

urben1680 commented on crates/bevy_ecs/src/relationship/related_methods.rs:
```diff
@@ -302,6 +303,44 @@ impl<'w> EntityWorldMut<'w> {
+    /// Despawns last entity that relate to this one via the given [`RelationshipTarget`].
+    /// This entity will not be despawned.
+    pub fn despawn_last_related<S: RelationshipTarget>(&mut self) -> &mut Self {
+        if let Some(last_child) = self.get::<Children>().and_then(|c| c.iter().last()) {
```
I am not sure this is the right approach as types like [`EntityHashSet` implement `RelationshipSourceCollection`](https://dev-docs.bevyengine.org/bevy/ecs/relationship/trait.RelationshipSourceCollection.html#impl-RelationshipSourceCollection-for-EntityHashSet) which' iteration order is unrelated to the insertion order, it maybe is even unstable.

You would have to introduce another trait like this:
```rs
pub trait InsertionOrderedRelationshipSourceCollection: RelationshipSourceCollection {}
```
... and implement it on the types that fit so you can use it as a bound on `S::Collection` here.

It might even make sense to give this new trait helper methods like `pop_last` and `pop_first` so no iterator is involved at all. Though maybe the names need to be different to not collide with existing ones, like in case `VecDeque` becomes a valid collection type, which may make sense if your new API is used often on larger collections.

MuongKimhong commented:
Hi @urben1680 

Thanks for pointing out the issue, I completely missed that detail.  
I've updated the code based on your feedback and introduced the `InsertionOrderedRelationshipSourceCollection` trait as suggested.

Please have a review.

alice-i-cecile commented:
Related to #20546.
